### PR TITLE
Mrossides/reverse segments

### PR DESCRIPTION
**Pull request by rossides**
First, we reverse the iteration of the segments.

In a follow up RB, we will reverse the edge iteration within a segment.

Staging in progress...

### CODE REVIEW

File: graphjet-core/src/main/java/com/twitter/graphjet/bipartite/MultiSegmentReverseIterator.java [hunk 1]
```diff
@@ -0,0 +1,58 @@
+/**
+ * Copyright 2016 Twitter. All rights reserved.
```
Comment (guimingTang) [inline graphjet-core/src/main/java/com/twitter/graphjet/bipartite/MultiSegmentReverseIterator.java]: 2017*

File: graphjet-core/src/main/java/com/twitter/graphjet/bipartite/MultiSegmentReverseIterator.java [hunk 2]
```diff
@@ -0,0 +1,58 @@
...
+
+import com.twitter.graphjet.bipartite.api.EdgeIterator;
+import com.twitter.graphjet.bipartite.segment.LeftIndexedBipartiteGraphSegment;
+
+/**
+ * This iterator provides reverse chronological access over edges where the edges can be spread across segments.
+ */
+public class MultiSegmentReverseIterator<T extends LeftIndexedBipartiteGraphSegment> extends MultiSegmentIterator<T> {
+  /**
+   * This constructor mirror the one in it's super-class to reuse common code.
```
Comment (guimingTang) [inline graphjet-core/src/main/java/com/twitter/graphjet/bipartite/MultiSegmentReverseIterator.java]: I like that you're adding context to clarify the usage, but I'm not sure if this is relevant to explain this reuses common code. Could you provide some context to what you're trying to explain? Also *its*

File: graphjet-core/src/main/java/com/twitter/graphjet/bipartite/MultiSegmentReverseIterator.java [hunk 3]
```diff
@@ -0,0 +1,58 @@
...
+    initializeCurrentSegmentIterator();
+    return this;
+  }
+
+  /**
+   * This finds segments in reverse chronological order.
+   * Returns false if it cannot find a non-empty next segment for the node
+   */
+  @Override
+  protected boolean findNextSegmentForNode() {
```
Comment (guimingTang) [inline graphjet-core/src/main/java/com/twitter/graphjet/bipartite/MultiSegmentReverseIterator.java]: I'm a bit concerned about overriding a method that is already implemented. For example, in this implementation:
MultiSegmentIterator. findNextSegmentForNode() 
vs 
MultiSegmentReverseIterator. findNextSegmentForNode() will result in 2 different behaviors. 
IMO it's better to super-class a MultiSegmentIterator and make it abstract, then sub-class 2 classes, one reverse-chronological, one chronological. What's your opinion on this?

Comment (jerryjiangabc) [inline graphjet-core/src/main/java/com/twitter/graphjet/bipartite/MultiSegmentReverseIterator.java]: second this.

File: graphjet-core/src/main/java/com/twitter/graphjet/bipartite/ChronologicalMultiSegmentIterator.java [hunk 4]
```diff
@@ -0,0 +1,56 @@
...
+ * limitations under the License.
+ */
+
+package com.twitter.graphjet.bipartite;
+
+import com.twitter.graphjet.bipartite.api.EdgeIterator;
+import com.twitter.graphjet.bipartite.segment.LeftIndexedBipartiteGraphSegment;
+
+/**
+ * This iterator provides reverse chronological access over edges where the edges can be spread across segments.
```
Comment (jerryjiangabc) [inline graphjet-core/src/main/java/com/twitter/graphjet/bipartite/ChronologicalMultiSegmentIterator.java]: reverse chronological => chronological

File: graphjet-core/src/main/java/com/twitter/graphjet/bipartite/MultiSegmentIterator.java [hunk 5]
```diff
@@ -65,67 +65,48 @@ private void rebuildSegmentIterators() {
     segmentEdgeAccessor.rebuildIterators(oldestSegmentId, liveSegmentId);
   }
 
-  @Override
-  public EdgeIterator resetForNode(long inputNode) {
+  public void rebuildSegmentIteratorsForNode(long inputNode) {
```
Comment (jerryjiangabc) [inline graphjet-core/src/main/java/com/twitter/graphjet/bipartite/MultiSegmentIterator.java]: public => protected

File: graphjet-core/src/main/java/com/twitter/graphjet/bipartite/MultiSegmentRandomIterator.java [hunk 6]
```diff
@@ -153,4 +147,8 @@ public byte currentEdgeType() {
   public boolean hasNext() {
     return numSamplesReturned < numSamplesNeeded;
   }
+
+  protected boolean findNextSegmentForNode() {
```
Comment (jerryjiangabc) [inline graphjet-core/src/main/java/com/twitter/graphjet/bipartite/MultiSegmentRandomIterator.java]: is this function needed? can we remove it?

File: graphjet-core/src/main/java/com/twitter/graphjet/bipartite/MultiSegmentRandomIterator.java [hunk 7]
```diff
@@ -121,9 +116,8 @@ private void sampleFromAliasTable() {
     }
   }
 
-  @Override
   public EdgeIterator resetForNode(long node, int numSamplesToGet, Random randomGen) {
-    super.resetForNode(node); // rebuilds the segmentEdgeRandomAccessor if needed
+    rebuildSegmentIteratorsForNode(node); // rebuilds the segmentEdgeRandomAccessor if needed
```
Comment (jerryjiangabc) [inline graphjet-core/src/main/java/com/twitter/graphjet/bipartite/MultiSegmentRandomIterator.java]: in the previous super.resetForNote(node), in addition to rebuildSegmentIteratorsForNode, there are other logics such as
currentSegmentId = oldestSegmentId;
initializeCurrentSegmentIterator();

do you intend to remove them?

File: graphjet-core/src/test/java/com/twitter/graphjet/bipartite/MultiSegmentPowerLawBipartiteGraphTest.java [hunk 8]
```diff
@@ -213,6 +213,21 @@ public void testMultiSegmentConstruction() throws Exception {
...
 
   @Test
+  public void testMultiSegmentReverseIteration() throws Exception {
+    NodeMetadataLeftIndexedPowerLawMultiSegmentBipartiteGraph nodeMetadataLeftIndexedPowerLawMultiSegmentBipartiteGraph =
+      new NodeMetadataLeftIndexedPowerLawMultiSegmentBipartiteGraph(
+        3, 3, 5, 2, 2.0, 5, 2, new IdentityEdgeTypeMask(), new NullStatsReceiver());
+
+    addEdges(nodeMetadataLeftIndexedPowerLawMultiSegmentBipartiteGraph);
+
+    assertEquals(new LongArrayList(new long[]{13}),
```
Comment (jerryjiangabc) [inline graphjet-core/src/test/java/com/twitter/graphjet/bipartite/MultiSegmentPowerLawBipartiteGraphTest.java]: A more complicated test case is to create a randomly generated graph and compare the results of two iterators, which should be the exactly same. in the test case, it is better to check edge order across different segments as well as edge order within the same segment.

**Pull request merged**